controllers/database_controller.go

```diff
@@ -38,153 +62,405 @@ func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 	if err != nil {
+		r.Log.Error(err, "adding permission to role failed", "role", readName, "permission", readPerms)
+		return err
+	}
+
+	writePerms := getDBWritePermissions(database)
+	err = createDefaultPermissions(stardogClient, auth, writeName, writePerms)
+	if err != nil {
+		r.Log.Error(err, "adding permission to role failed", "role", writeName, "permission", writePerms)
+		return err
+	}
+
+	// assign roles to users
+	readUserRolesResp, err := stardogClient.UsersRoles.ListUserRoles(users_roles.NewListUserRolesParams().WithUser(readName), auth)
+	if err != nil || !readUserRolesResp.IsSuccess() {
 		r.Log.Error(err, "error getting user roles", "user", readName)
-		return ctrl.Result{}, err
+		return err
 	}
 
-	if len(readUserRoles) != 1 || readUserRoles[0] != readName {
-		err = apiClient.SetUserRoles(ctx, readName, []string{readName})
-		if err != nil {
-			r.Log.Error(err, "error assigning roles to user", "user", readName, "roles", readUserRoles)
-			return ctrl.Result{}, err
+	if !slices.Contains(readUserRolesResp.Payload.Roles, readName) {
+		params := users_roles.NewAddRoleParams().
+			WithUser(readName).
+			WithRole(&models.Rolename{Rolename: &readName})
+		roleResp, err := stardogClient.UsersRoles.AddRole(params, auth)
+		if err != nil || !roleResp.IsSuccess() {
+			r.Log.Error(err, "error assigning role to user", "user", readName, "role", readName)
+			return err
 		}
 	}
 
-	writeUserRoles, err := apiClient.GetUserRoles(ctx, writeName)
-	if err != nil {
+	writeUserRolesResp, err := stardogClient.UsersRoles.ListUserRoles(users_roles.NewListUserRolesParams().WithUser(writeName), auth)
+	if err != nil || !writeUserRolesResp.IsSuccess() {
 		r.Log.Error(err, "error getting user roles", "user", writeName)
-		return ctrl.Result{}, err
+		return err
 	}
 
-	if len(writeUserRoles) != 2 || writeUserRoles[0] != readName || writeUserRoles[1] != writeName {
-		err = apiClient.SetUserRoles(ctx, writeName, []string{readName, writeName})
-		if err != nil {
-			r.Log.Error(err, "error assigning roles to user", "user", readName, "roles", readUserRoles)
-			return ctrl.Result{}, err
+	if !slices.Contains(writeUserRolesResp.Payload.Roles, readName) {
+		params := users_roles.NewAddRoleParams().
+			WithUser(writeName).
+			WithRole(&models.Rolename{Rolename: &readName})
+		roleResp, err := stardogClient.UsersRoles.AddRole(params, auth)
+		if err != nil || !roleResp.IsSuccess() {
+			r.Log.Error(err, "error assigning role to user", "user", writeName, "role", readName)
+			return err
+		}
+	}
+
+	if !slices.Contains(writeUserRolesResp.Payload.Roles, writeName) {
+		params := users_roles.NewAddRoleParams().
+			WithUser(writeName).
+			WithRole(&models.Rolename{Rolename: &writeName})
+		roleResp, err := stardogClient.UsersRoles.AddRole(params, auth)
+		if err != nil || !roleResp.IsSuccess() {
+			r.Log.Error(err, "error assigning role to user", "user", writeName, "role", writeName)
+			return err
 		}
```

Comment:
Nitpicky Nitpick (feel free to ignore): I'd probably move the logic to create user roles to a helper function `assignDefaultRoles` or something. The same way you did for the permissions. 

You're mixing layers here which makes it a bit harder to read.